corehq/apps/callcenter/utils.py

```diff
@@ -185,22 +172,28 @@ def valid_element_name(name):
         'phone_number': commcare_user.phone_number or '',
         'last_device_id_used': commcare_user.devices[0].device_id if commcare_user.devices else '',
     })
+    fields.pop('case_type', None)
 
     return fields
 
 
-def _user_case_changed(case, case_type, close, fields, owner_id):
+def _get_changed_fields(case, case_type, owner_id, fields):
+    changed_fields = {}
+    if case.name != fields['name']:
+        changed_fields['name'] = fields['name']
+
+    if case.type != case_type:
+        changed_fields['case_type'] = case_type
+
+    if case.owner_id != owner_id:
+        changed_fields['owner_id'] = owner_id
+
     props = case.dynamic_case_properties()
-    changed = close and not case.closed
-    changed = changed or case.type != case_type
-    changed = changed or case.name != fields['name']
-    changed = changed or case.owner_id != owner_id
-    if not changed:
-        for field, value in fields.items():
-            if field != 'name' and props.get(field) != value:
-                changed = True
-                break
-    return changed
+    for field, value in fields.items():
+        if field != 'name' and props.get(field) != value:
+            changed_fields[field] = value
```

Comment:
This will overwrite `case_type` and `owner_id` with values from `fields` (if they exist) even if they were set above on lines 186 and 189. Is that intentional? (that's why I suggested using `setdefault`)

As is, I think this function has confusing semantics. It is not obvious what will be returned when called like
```py
changed = _get_changed_fields(case, "type_one", "bob", {
    "case_type": "type_two",
    "owner_id": "frank",
})
```
Maybe it should throw an exception if `"case_type"` or `"owner_id"` are present in `fields`? Or maybe those arguments should be removed from this function so the only way to set them is by passing them in the `fields` dict (this could result in a cleaner implementation too, both here and above in `sync_user_case`)?

Reply:
`case_type` and `owner_id` should never be in fields (though that's not obvious from the context of this function.

I've made and update to add `case_type` and `owner_id` into fields which simplifies things somewhat